File: example_py/basics3_self_collision_avoidance.py
#!/usr/bin/env python

"""basics3_self_collision_avoidance.py

This tutorial shows how to use the self collision monitor to keep the left and right robot from
hitting each other.
"""

__copyright__ = "Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse
import spdlog
import flexivdrdk
from flexivrdk import Mode  # for type definition


def main():
    # Program Setup
    # ==============================================================================================
    # Parse arguments
    argparser = argparse.ArgumentParser()
    argparser.add_argument(
        "left_robot_sn",
        help="Serial number of the left robot to connect. Remove any space, e.g. Rizon4s-100001",
    )
    argparser.add_argument(
        "right_robot_sn",
        help="Serial number of the right robot to connect. Remove any space, e.g. Rizon4s-100002",
    )
    argparser.add_argument(
        "left_robot_urdf",
        help="Absolute path to the URDF of the left robot",
    )
    argparser.add_argument(
        "right_robot_urdf",
        help="Absolute path to the URDF of the right robot",
    )
    args = argparser.parse_args()

    # Define alias
    logger = spdlog.ConsoleLogger("Example")

    # Print description
    logger.info(
        ">>> Tutorial description <<<\nThis tutorial shows how to use the self collision monitor "
        "to keep the left and right robot from hitting each other.\n"
    )

    try:
        # DRDK Initialization
        # ==========================================================================================
        # Translation in world of the left and robot
        left_trans_in_world = [0, 0.3, 0]
        right_trans_in_world = [0, -0.3, 0]

        # Instantiate robot pair interface
        robot_pair = flexivdrdk.RobotPair(
            [args.left_robot_sn, args.right_robot_sn],
            [left_trans_in_world, right_trans_in_world],
        )

        # Self Collision Monitor
        # ==========================================================================================
        # Instantiate self collision monitor interface
        coll_monitor = flexivdrdk.SelfCollisionMonitor(
            robot_pair, [args.left_robot_urdf, args.right_robot_urdf]
        )

        # Use 10 cm safety distance, leaving some buffer for the robots to decelerate
        coll_monitor.SetMinDistance(0.1)
        logger.info("Safety distance set to 10 cm")

        # Start collision monitor and keep it running in the background
        coll_monitor.Start()
        logger.info("Self-collision monitor started")

        # Move robots to home pose
        robot_pair.SwitchMode(Mode.NRT_PLAN_EXECUTION)
        robot_pair.ExecutePlan(["PLAN-Home", "PLAN-Home"])
        while robot_pair.busy():
            time.sleep(1)
        logger.info("Robot pair reached home")

        # Move the two robots to the same Cartesian location in world
        logger.info("Moving the robots towards each other")
        robot_pair.SwitchMode(Mode.NRT_CARTESIAN_MOTION_FORCE)
        target_pose = [0.5, 0.0, 0.3, 0, 0, 1, 0]
        robot_pair.SendCartesianMotionForce([target_pose, target_pose])

        # Both robots will stop before they hit each other
        time.sleep(1)
        while not robot_pair.stopped():
            time.sleep(1)

        logger.info("Program finished")

    except Exception as e:
        # Print exception error message
        logger.error(str(e))


if __name__ == "__main__":
    main()

File: example/basics3_self_collision_avoidance.cpp
/**
 * @example basics3_self_collision_avoidance.cpp
 * This tutorial shows how to use the self collision monitor to keep the left and right robot from
 * hitting each other.
 * @copyright Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/drdk/robot_pair.hpp>
#include <flexiv/drdk/self_collision_monitor.hpp>
#include <flexiv/rdk/utility.hpp>

#include <iostream>
#include <string>
#include <thread>

using namespace flexiv;

/** @brief Print program usage help */
void PrintHelp()
{
    // clang-format off
    std::cout << "Required arguments: [left_robot_sn] [right_robot_sn] [left_robot_urdf] [right_robot_urdf]" << std::endl;
    std::cout << "    left_robot_sn: Serial number of the left robot to connect. Remove any space, e.g. Rizon4s-100001" << std::endl;
    std::cout << "    right_robot_sn: Serial number of the right robot to connect. Remove any space, e.g. Rizon4s-100002" << std::endl;
    std::cout << "    left_robot_urdf: Absolute path to the URDF of the left robot" << std::endl;
    std::cout << "    right_robot_urdf: Absolute path to the URDF of the right robot" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

int main(int argc, char* argv[])
{
    // Program Setup
    // =============================================================================================
    // Parse parameters
    if (argc < 5 || rdk::utility::ProgramArgsExistAny(argc, argv, {"-h", "--help"})) {
        PrintHelp();
        return 1;
    }

    // Serial numbers of the left and right robot respectively
    std::string left_robot_sn = argv[1];
    std::string right_robot_sn = argv[2];

    // URDF paths of the left and right robot respectively
    std::string left_robot_urdf = argv[3];
    std::string right_robot_urdf = argv[4];

    // Print description
    std::cout << ">>> Tutorial description <<<\nThis tutorial shows how to use the self collision "
                 "monitor to keep the left and right robot from hitting each other.\n"
              << std::endl;

    try {
        // DRDK Initialization
        // =========================================================================================
        // Translation in world of the left and robot
        std::array<double, 3> left_trans_in_world = {0, 0.3, 0};
        std::array<double, 3> right_trans_in_world = {0, -0.3, 0};

        // Instantiate robot pair interface
        drdk::RobotPair robot_pair(
            {left_robot_sn, right_robot_sn}, {left_trans_in_world, right_trans_in_world});

        // Self Collision Monitor
        // =========================================================================================
        // Instantiate self collision monitor interface
        drdk::SelfCollisionMonitor coll_monitor(robot_pair, {left_robot_urdf, right_robot_urdf});

        // Use 10 cm safety distance, leaving some buffer for the robots to decelerate
        coll_monitor.SetMinDistance(0.1);
        std::cout << "Safety distance set to 10 cm" << std::endl;

        // Start collision monitor and keep it running in the background
        coll_monitor.Start();
        std::cout << "Self-collision monitor started" << std::endl;

        // Move the two robots to home pose
        robot_pair.SwitchMode(rdk::Mode::NRT_PLAN_EXECUTION);
        robot_pair.ExecutePlan({"PLAN-Home", "PLAN-Home"});
        do {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        } while (robot_pair.busy());
        std::cout << "Robot pair reached home" << std::endl;

        // Move the two robots to the same Cartesian location in world
        std::cout << "Moving the robots towards each other" << std::endl;
        robot_pair.SwitchMode(rdk::Mode::NRT_CARTESIAN_MOTION_FORCE);
        std::array<double, rdk::kPoseSize> target_pose = {0.5, 0.0, 0.3, 0, 0, 1, 0};
        robot_pair.SendCartesianMotionForce({target_pose, target_pose});

        // Both robots will stop before they hit each other
        do {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        } while (!robot_pair.stopped());

        std::cout << "Program finished" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "[error] " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
